log4j-api/src/main/java/org/apache/logging/log4j/spi/LoggingSystem.java

```diff
@@ -0,0 +1,349 @@
+            this.provider = provider;
+        }
+
+        public LoggerContextFactory createLoggerContextFactory() {
+            final PropertyEnvironment environment = PropertiesUtil.getProperties();
+            final String customFactoryClass = environment.getStringProperty(LogManager.FACTORY_PROPERTY_NAME);
+            if (customFactoryClass != null) {
+                final LoggerContextFactory customFactory = createCustomFactory(customFactoryClass);
+                if (customFactory != null) {
+                    return customFactory;
+                }
+            }
+            if (provider != null) {
+                final Class<? extends LoggerContextFactory> factoryClass = provider.loadLoggerContextFactory();
+                if (factoryClass != null) {
+                    final LoggerContextFactory factory = tryInstantiate(factoryClass);
+                    if (factory != null) {
+                        return factory;
+                    }
+                }
+            }
+            LowLevelLogUtil.log("Log4j could not find a logging implementation. " +
+                    "Please add log4j-core dependencies to classpath or module path. " +
+                    "Using SimpleLogger to log to the console.");
+            return SimpleLoggerContextFactory.INSTANCE;
+        }
+
+        /**
+         * Creates the ThreadContextMap instance used by the ThreadContext.
+         * <p>
+         * If {@linkplain Constants#isThreadLocalsEnabled() Log4j can use ThreadLocals}, a garbage-free StringMap-based context map can
+         * be installed by setting system property {@code log4j2.garbagefree.threadContextMap} to {@code true}.
+         * </p><p>
+         * Furthermore, any custom {@code ThreadContextMap} can be installed by setting system property
+         * {@code log4j2.threadContextMap} to the fully qualified class name of the class implementing the
+         * {@code ThreadContextMap} interface. (Also implement the {@code ReadOnlyThreadContextMap} interface if your custom
+         * {@code ThreadContextMap} implementation should be accessible to applications via the
+         * {@link ThreadContext#getThreadContextMap()} method.)
+         * </p><p>
+         * Instead of system properties, the above can also be specified in a properties file named
+         * {@code log4j2.component.properties} in the classpath.
+         * </p>
+         *
+         * @see ThreadContextMap
+         * @see ReadOnlyThreadContextMap
+         * @see org.apache.logging.log4j.ThreadContext
+         */
+        public ThreadContextMap createContextMap() {
+            final PropertyEnvironment environment = PropertiesUtil.getProperties();
+            // TODO(ms): make property lookup lazier here to avoid eager init
+            GarbageFreeSortedArrayThreadContextMap.init(environment);
+            CopyOnWriteSortedArrayThreadContextMap.init(environment);
+            DefaultThreadContextMap.init(environment);
```

Comment:
These used to be initialized by `ThreadContextMapFactory::init`, but they're here now. Ideally, those classes should use a `PropertyEnvironment` constructor argument or similar, though I've deferred refactoring all of that until a later commit as this one was already getting big enough.

Reply:
Fixing this issue in a subsequent commit.